File: backend/src/main/scala/services/battle/microservices/combat/services/BattleProjectileFactoryRules.scala
package services.battle.microservices.combat.services

import cats.effect.IO

import services.battle.microservices.world.services.BattleGeometry.*
import services.battle.microservices.combat.objects.combat.BattleWeaponProjectileDefinition
import services.battle.objects.core.{BattleAggregateState, BattleVector2, ClientCommandSeq, FacingRadians, ProjectileId}
import services.battle.microservices.actors.objects.player.BattlePlayerState
import services.battle.microservices.combat.objects.projectile.BattleProjectileState

private[battle] object BattleProjectileFactoryRules {
  def weaponProjectiles(
    shooter: BattlePlayerState,
    commandSeq: ClientCommandSeq,
    projectile: BattleWeaponProjectileDefinition,
    projectileBirthOffset: Double,
    normalizeMovement: BattleVector2 => IO[BattleVector2]
  ): IO[Vector[BattleProjectileState]] = {
    val projectileCount = math.max(1, projectile.projectileCount.value)
    (0 until projectileCount).toVector.foldLeft(IO.pure(Vector.empty[BattleProjectileState])) { (projectilesIO, index) =>
      for
        projectiles <- projectilesIO
        direction <- spreadDirection(shooter.aim, index, projectileCount, projectile.spread.value, normalizeMovement)
        id <- projectileId(shooter, commandSeq, index, projectileCount)
        position <- projectileBirthPosition(shooter, direction, projectileBirthOffset, normalizeMovement)
        nextProjectile <- weaponProjectile(shooter, id, position, direction, projectile)
      yield projectiles :+ nextProjectile
    }
  }

  def resolvePistolShot(
    state: BattleAggregateState,
    shooter: BattlePlayerState,
    commandSeq: ClientCommandSeq,
    projectile: BattleWeaponProjectileDefinition,
    projectileBirthOffset: Double,
    normalizeMovement: BattleVector2 => IO[BattleVector2]
  ): IO[BattleAggregateState] = {
    val direction = shooter.aim
    for
      start <- projectileBirthPosition(shooter, direction, projectileBirthOffset, normalizeMovement)
      projectileState <- pistolProjectile(shooter, commandSeq, start, direction, projectile)
    yield state.copy(projectiles = state.projectiles :+ projectileState)
  }

  private def projectileId(
    shooter: BattlePlayerState,
    commandSeq: ClientCommandSeq,
    projectileIndex: Int,
    projectileCount: Int
  ): IO[ProjectileId] = IO.pure {
    val suffix = if projectileCount == 1 then "" else s"-$projectileIndex"
    ProjectileId(s"projectile-${shooter.playerId.value}-${commandSeq.value}$suffix")
  }

  private def spreadDirection(
    direction: BattleVector2,
    projectileIndex: Int,
    projectileCount: Int,
    spreadRadians: Double,
    normalizeMovement: BattleVector2 => IO[BattleVector2]
  ): IO[BattleVector2] =
    if projectileCount <= 1 || spreadRadians == 0.0 then IO.pure(direction)
    else
      val offset =
        ((projectileIndex.toDouble / (projectileCount - 1).toDouble) - 0.5) * spreadRadians
      rotate(direction, offset, normalizeMovement)

  private def rotate(
    direction: BattleVector2,
    radians: Double,
    normalizeMovement: BattleVector2 => IO[BattleVector2]
  ): IO[BattleVector2] = {
    val cos = math.cos(radians)
    val sin = math.sin(radians)
    normalizeMovement(
      BattleVector2(
        direction.x * cos - direction.y * sin,
        direction.x * sin + direction.y * cos
      )
    )
  }

  private def weaponProjectile(
    shooter: BattlePlayerState,
    projectileId: ProjectileId,
    position: BattleVector2,
    direction: BattleVector2,
    projectile: BattleWeaponProjectileDefinition
  ): IO[BattleProjectileState] =
    scale(direction, projectile.speed.value).map { velocity =>
      BattleProjectileState(
        projectileId = projectileId,
        ownerHeroId = shooter.heroId,
        projectileKind = projectile.projectileKind,
        position = position,
        velocity = velocity,
        facing = FacingRadians(math.atan2(direction.y, direction.x)),
        radius = projectile.radius,
        damage = projectile.damage,
        ttlMs = projectile.lifetime,
        maxLifetimeMs = projectile.lifetime,
        splashRadius = projectile.splashRadius
      )
    }

  private def pistolProjectile(
    shooter: BattlePlayerState,
    commandSeq: ClientCommandSeq,
    start: BattleVector2,
    direction: BattleVector2,
    projectile: BattleWeaponProjectileDefinition
  ): IO[BattleProjectileState] =
    scale(direction, projectile.speed.value).map { velocity =>
      BattleProjectileState(
        projectileId = ProjectileId(s"projectile-${shooter.playerId.value}-${commandSeq.value}"),
        ownerHeroId = shooter.heroId,
        projectileKind = projectile.projectileKind,
        position = start,
        velocity = velocity,
        facing = shooter.facing,
        radius = projectile.radius,
        damage = projectile.damage,
        ttlMs = projectile.lifetime,
        maxLifetimeMs = projectile.lifetime,
        splashRadius = projectile.splashRadius
      )
    }

  private def projectileBirthPosition(
    shooter: BattlePlayerState,
    direction: BattleVector2,
    projectileBirthOffset: Double,
    normalizeMovement: BattleVector2 => IO[BattleVector2]
  ): IO[BattleVector2] =
    for
      normalized <- normalizeMovement(direction)
      offset <- scale(normalized, projectileBirthOffset)
      position <- add(shooter.position, offset)
    yield position
}

File: frontend/src/runtime/battle/microservices/combat/functions/BattleProjectileFactoryRules.ts
import type { BattleHeroViewState as Hero } from "../../../../../objects/battle/microservices/actors/objects/player/BattleHeroViewState";
import type { BattleProjectileState as Projectile } from "../../../../../objects/battle/microservices/combat/objects/projectile/BattleProjectileState";
import type { BattleWeaponRuleDefinition as WeaponDefinition } from "../../../../../objects/battle/microservices/combat/objects/combat/BattleCombatRuleDefinitions";
import type { BattleVector2 as Vec2 } from "../../../../../objects/battle/objects/core/BattleCoreScalars";

export const AUTHORITATIVE_PROJECTILE_BIRTH_CLEARANCE = 4;

export interface BattleWeaponProjectileAnglePlanInput {
  aimAngle: number;
  weaponDefinition: Pick<WeaponDefinition, "projectileCount" | "spreadRadians">;
}

export interface ProjectileBirthPositionInput {
  ownerPosition: Vec2;
  direction: Vec2;
  ownerRadius: number;
  projectileRadius: number;
}

export interface BattleProjectileSpawnInput {
  projectileSequence: number;
  player: Pick<Hero, "heroId" | "team" | "position" | "radius">;
  definition: WeaponDefinition;
  angle: number;
}

export function buildWeaponProjectileAnglePlan(input: BattleWeaponProjectileAnglePlanInput): number[] {
  const projectileCount = Math.max(1, Math.floor(input.weaponDefinition.projectileCount));
  if (projectileCount <= 1 || input.weaponDefinition.spreadRadians === 0) {
    return [input.aimAngle];
  }

  return Array.from({ length: projectileCount }, (_, projectileIndex) => {
    const offset = (projectileIndex / (projectileCount - 1) - 0.5) * input.weaponDefinition.spreadRadians;
    return input.aimAngle + offset;
  });
}

export function resolveProjectileBirthForwardDistance(ownerRadius: number, projectileRadius: number): number {
  return ownerRadius + projectileRadius + AUTHORITATIVE_PROJECTILE_BIRTH_CLEARANCE;
}

export function resolveProjectileBirthPosition(input: ProjectileBirthPositionInput): Vec2 {
  const forwardDistance = resolveProjectileBirthForwardDistance(input.ownerRadius, input.projectileRadius);
  return {
    x: input.ownerPosition.x + input.direction.x * forwardDistance,
    y: input.ownerPosition.y + input.direction.y * forwardDistance
  };
}

export function createBattleProjectileSpawn(input: BattleProjectileSpawnInput): Projectile {
  const direction = {
    x: Math.cos(input.angle),
    y: Math.sin(input.angle)
  };
  const position = resolveProjectileBirthPosition({
    ownerPosition: input.player.position,
    direction,
    ownerRadius: input.player.radius,
    projectileRadius: input.definition.projectileRadius
  });

  return {
    projectileId: `projectile-${input.projectileSequence}`,
    kind: input.definition.projectileKind,
    ownerHeroId: input.player.heroId,
    team: input.player.team,
    position,
    velocity: {
      x: direction.x * input.definition.projectileSpeedPerSecond,
      y: direction.y * input.definition.projectileSpeedPerSecond
    },
    facing: input.angle,
    radius: input.definition.projectileRadius,
    damage: input.definition.projectileDamage,
    ttlMs: input.definition.projectileLifetimeMs,
    maxLifetimeMs: input.definition.projectileLifetimeMs,
    splashRadius: input.definition.splashRadius,
    alive: true,
    hitTargets: []
  };
}
